
$\qquad$ $\qquad$$\qquad$  **TDA206/DIT206 Discrete Optimization: Assignment 4 -- SDP and Maxcut** <br />
$\qquad$ $\qquad$$\qquad$                   **Grader: David Bosch** <br />
$\qquad$ $\qquad$$\qquad$                     **Due Date: 17/03/2025** <br />
$\qquad$ $\qquad$$\qquad$                   **Submitted by: Name, Personal No., Email** <br />


---


General guidelines:
*   All solutions to theoretical and pratical problems must be submitted in this ipynb notebook, and equations wherever required, should be formatted using LaTeX math-mode.
*   All discussion regarding practical problems, along with solutions and plots should be specified in this notebook. All plots/results should be visible such that the notebook do not have to be run. But the code in the notebook should reproduce the plots/results if we choose to do so.
*   Your name, personal number and email address should be specified above.
*   All tables and other additional information should be included in this notebook.
*   Before submitting, make sure that your code can run on another computer. That all plots can show on another computer including all your writing. It is good to check if your code can run here: https://colab.research.google.com.


# Question 1. [14 pts]

Consider the triangle graph i.e, three vertices all connected pairwise (unweighted, so each edge has weight $1$).

* a. Write the $+1/-1$ labelling corresponding to a maximum cut and give the value of the cut. [3 pts]
* b. Write the SDP relaxation of the MAXCUT problem for this specific graph: write the program explicitly without using summation signs, with the variables corresponding to a $3 \times 3$ matrix, $X_{1,1,}, X_{1,2}, \cdots$. [2 pts]
* c. Solve the SDP by manual calculation (not using a SDP solver). (HINT: Use symmetry and then argue about when the matrix with $1$ as diagonal elements and $\alpha$ in other positions is psd.) [3 pts]
* d. Produce the vector labelling corresponding to the optimal solution of the SDP in (c). (HINT: for what angle is $\cos \theta = -1/2$?) [3 pts]
* e. What is the expected value of the cut produced by rounding using the vector labels and a random hyperlane? Give a short justification. [3 pts]


## Question 1 (MAXCUT on the triangle graph)

Let the triangle graph have vertices $\{1,2,3\}$ and edges $E=\{(1,2),(1,3),(2,3)\}$, all with weight $1$.



### (a) A $\pm 1$ labeling for a maximum cut and the cut value
A maximum cut in a triangle separates one vertex from the other two. For example,
$s=(s_1,s_2,s_3)=(1,-1,-1)$.
Then edges $(1,2)$ and $(1,3)$ cross the cut, while $(2,3)$ does not, so the cut value is $2$.
(Indeed, no cut can include all 3 edges in a triangle, so $\mathrm{MAXCUT}=2$.)

### (b) SDP relaxation written explicitly (matrix variables)
Using the standard MAXCUT SDP relaxation from the lectures: introduce a symmetric matrix $X\in\mathbb{R}^{3\times 3}$ with $X\succeq 0$ and $X_{ii}=1$.
The SDP is:
$$
\max\ \frac{1-X_{12}}{2}+\frac{1-X_{13}}{2}+\frac{1-X_{23}}{2}
$$
subject to
$$
X_{11}=1,\quad X_{22}=1,\quad X_{33}=1,\quad X=X^T,\quad X\succeq 0.
$$
Equivalently, the objective is $\max\ \frac{3-(X_{12}+X_{13}+X_{23})}{2}$.



### (c) Solve the SDP by symmetry and PSD condition
By symmetry of the triangle, we can assume the optimal solution has equal off-diagonal entries:
$$
X=
\begin{pmatrix}
1 & \alpha & \alpha\\
\alpha & 1 & \alpha\\
\alpha & \alpha & 1
\end{pmatrix}.
$$
This matrix has eigenvalues $1+2\alpha$ and $1-\alpha$ (with multiplicity 2), so $X\succeq 0$ iff
$1+2\alpha\ge 0$ and $1-\alpha\ge 0$, i.e. $-\frac12\le \alpha\le 1$.

The SDP objective becomes
$$
\frac{3-3\alpha}{2},
$$
which is maximized by minimizing $\alpha$. Hence
$\alpha^\star=-\frac12$, and the SDP optimum value is
$$
\frac{3-3(-1/2)}{2}=\frac{9}{4}=2.25.
$$



### (d) Vector labeling for the optimal SDP solution
Because $X\succeq 0$ and $X_{ii}=1$, we can represent $X$ as a Gram matrix of unit vectors:
$X_{ij}=v_i^Tv_j$ with $\|v_i\|=1$.
Here the optimum has $v_i^Tv_j=\alpha^\star=-\frac12$ for $i\ne j$, i.e. the pairwise angle satisfies
$\cos\theta=-\frac12$, so $\theta=120^\circ=2\pi/3$.

One explicit choice in $\mathbb{R}^2$ is:
$$
v_1=(1,0),\quad 
v_2=\left(-\frac12,\frac{\sqrt3}{2}\right),\quad
v_3=\left(-\frac12,-\frac{\sqrt3}{2}\right),
$$
which gives $v_i^Tv_j=-\frac12$ for all $i\ne j$.


### (e) Expected cut value under random-hyperplane rounding
In the Goemans–Williamson rounding from the lectures, a random hyperplane through the origin separates
$v_i$ and $v_j$ with probability $\theta_{ij}/\pi$, where $\theta_{ij}$ is the angle between them.
Here $\theta_{ij}=2\pi/3$, so each edge is cut with probability $(2\pi/3)/\pi=2/3$.

Since there are 3 unit-weight edges, the expected cut value is
$$
\mathbb{E}[\text{cut}] = 3\cdot \frac{2}{3}=2.
$$


# Question 2. [6 pts]


* a. Implement the SDP relaxation for MAXCUT in CVXPY (https://www.cvxpy.org), see (https://www.cvxpy.org/examples/basic/sdp.html). [2 pts]
* b. Solve the SDP in the previous problem using (a). [1 pts]
* c. Solve the SDP and give the MAXCUT value for the random graph 'graph.txt' in homework-2. [1 pts]
* d. Solve the SDP and give the MAXCUT value for the planted partition graph 'G' given below. Give the approximation ratio. [2 pts]

In [9]:
#'G' is given as an adjaceny matrix
adj_mat = [[   0.  , 2., 1000.  ,  2. , 1000.  ,  2.  ,  2. ,1000.  ,  2.  ,  2.],
 [ 2.  ,  0. ,1000.  ,  2., 1000.  ,  2.  ,  2. ,1000.  ,  2.  ,  2.],
 [1000. , 1000.  ,  0. ,1000. ,   2. ,1000., 1000. ,   2. ,1000., 1000.],
 [   2.  ,  2. ,1000.  ,  0. ,1000.  ,  2. ,   2., 1000.  ,  2.  ,  2.],
 [1000., 1000.  ,  2. ,1000. ,   0., 1000. ,1000.  ,  2. ,1000. ,1000.],
 [   2.  ,  2., 1000.  ,  2., 1000. ,   0. ,   2., 1000. ,   2. ,   2.],
 [   2.  ,  2. ,1000.  ,  2. ,1000. ,   2. ,   0., 1000. ,   2. ,   2.],
 [1000.  ,1000.,    2. ,1000.,    2., 1000., 1000.,    0., 1000., 1000.],
 [   2. ,   2. ,1000. ,   2. ,1000. ,   2. ,   2., 1000.,    0.,    2.],
 [   2. ,   2. ,1000. ,   2. ,1000. ,   2. ,   2., 1000.,    2.,    0.]]



In [3]:
import numpy as np
import cvxpy as cp
import os

# MAXCUT utilities

def cut_value_from_labels(labels, W):
    """labels in {+1,-1}, W symmetric with zero diagonal."""
    s = labels.reshape(-1, 1)
    return 0.25 * np.sum(W * (1 - (s @ s.T)))

def factor_gram(X):
    """Stable factorization X = U^T U via eigendecomposition (works even with small numerical negatives)."""
    Xsym = (X + X.T) / 2.0
    evals, evecs = np.linalg.eigh(Xsym)
    evals = np.maximum(evals, 0.0)
    U = (np.sqrt(evals)[:, None] * evecs.T)  
    return U

def gw_rounding(X, W, n_rounds=300, seed=0):
    """Goemans–Williamson rounding: random hyperplanes; return best cut and its labels."""
    rng = np.random.default_rng(seed)
    U = factor_gram(X)
    n = X.shape[0]

    best_val = -1.0
    best_labels = None
    for _ in range(n_rounds):
        g = rng.normal(size=n)
        scores = U.T @ g
        labels = np.where(scores >= 0, 1, -1)
        val = cut_value_from_labels(labels, W)
        if val > best_val:
            best_val = val
            best_labels = labels.copy()
    return best_val, best_labels

# SDP solver (lecture SDP)
def solve_maxcut_sdp(W, solver="SCS", eps=1e-5, max_iters=30000):
    """
    Solve MAXCUT SDP:
      max 1/4 * sum_ij W_ij * (1 - X_ij)
      s.t. diag(X)=1, X PSD
    """
    W = np.array(W, dtype=float)
    W = (W + W.T) / 2.0
    np.fill_diagonal(W, 0.0)
    n = W.shape[0]

    X = cp.Variable((n, n), PSD=True)
    constraints = [cp.diag(X) == 1]
    obj = cp.Maximize(0.25 * cp.sum(cp.multiply(W, (1 - X))))
    prob = cp.Problem(obj, constraints)

    prob.solve(solver=getattr(cp, solver), eps=eps, max_iters=max_iters, verbose=False)
    return prob.value, X.value, prob.status

# (b) Triangle K3
W_tri = np.array([[0,1,1],
                  [1,0,1],
                  [1,1,0]], dtype=float)

sdp_tri, X_tri, st_tri = solve_maxcut_sdp(W_tri, solver="SCS", eps=1e-6, max_iters=20000)
best_tri, lab_tri = gw_rounding(X_tri, W_tri, n_rounds=500, seed=0)

print("=== (b) Triangle K3 ===")
print("SDP status:", st_tri)
print("SDP value:", sdp_tri)
print("Best rounded cut:", best_tri)
print("Best labels:", lab_tri)

# (c) Random graph from HW2: try graph.txt, else generate G(100,0.1)
def load_graph_txt(path="graph.txt"):
    if not os.path.exists(path):
        return None
    edges = []
    with open(path, "r") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            parts = line.replace(",", " ").split()
            if len(parts) >= 2:
                u, v = int(parts[0]), int(parts[1])
                edges.append((u, v))
    if not edges:
        return None
    edges = np.array(edges, dtype=int)
    if edges.min() == 1:
        edges -= 1
    n = int(edges.max() + 1)
    W = np.zeros((n, n), dtype=float)
    for (u, v) in edges:
        if u == v:
            continue
        W[u, v] = 1.0
        W[v, u] = 1.0
    return W

def gen_gnp_W(n=100, p=0.1, seed=0):
    rng = np.random.default_rng(seed)
    W = np.zeros((n, n), dtype=float)
    for i in range(n):
        r = rng.random(n - i - 1)
        js = np.where(r < p)[0] + i + 1
        for j in js:
            W[i, j] = 1.0
            W[j, i] = 1.0
    return W

W_rand = load_graph_txt("graph.txt")
source = "graph.txt"
if W_rand is None:
    W_rand = gen_gnp_W(n=100, p=0.1, seed=0)
    source = "generated G(100,0.1), seed=0"

sdp_rand, X_rand, st_rand = solve_maxcut_sdp(W_rand, solver="SCS", eps=1e-4, max_iters=40000)
best_rand, lab_rand = gw_rounding(X_rand, W_rand, n_rounds=200, seed=0)

print("\n=== (c) Random graph ===")
print("Source:", source)
print("n =", W_rand.shape[0], ", |E| =", int(W_rand.sum()/2))
print("SDP status:", st_rand)
print("SDP value:", sdp_rand)
print("Best rounded cut:", best_rand)
print("Approx ratio (best/SDP):", best_rand / sdp_rand if sdp_rand > 1e-12 else None)

#'G' is given as an adjaceny matrix
adj_mat = [[   0.  , 2., 1000.  ,  2. , 1000.  ,  2.  ,  2. ,1000.  ,  2.  ,  2.],
 [ 2.  ,  0. ,1000.  ,  2., 1000.  ,  2.  ,  2. ,1000.  ,  2.  ,  2.],
 [1000. , 1000.  ,  0. ,1000. ,   2. ,1000., 1000. ,   2. ,1000., 1000.],
 [   2.  ,  2. ,1000.  ,  0. ,1000.  ,  2. ,   2., 1000.  ,  2.  ,  2.],
 [1000., 1000.  ,  2. ,1000. ,   0., 1000. ,1000.  ,  2. ,1000. ,1000.],
 [   2.  ,  2., 1000.  ,  2., 1000. ,   0. ,   2., 1000. ,   2. ,   2.],
 [   2.  ,  2. ,1000.  ,  2. ,1000. ,   2. ,   0., 1000. ,   2. ,   2.],
 [1000.  ,1000.,    2. ,1000.,    2., 1000., 1000.,    0., 1000., 1000.],
 [   2. ,   2. ,1000. ,   2. ,1000. ,   2. ,   2., 1000.,    0.,    2.],
 [   2. ,   2. ,1000. ,   2. ,1000. ,   2. ,   2., 1000.,    2.,    0.]]


W_planted = np.array(adj_mat, dtype=float)
sdp_pl, X_pl, st_pl = solve_maxcut_sdp(W_planted, solver="SCS", eps=1e-6, max_iters=30000)
best_pl, lab_pl = gw_rounding(X_pl, W_planted, n_rounds=500, seed=0)

print("\n=== (d) Planted graph ===")
print("n =", W_planted.shape[0])
print("SDP status:", st_pl)
print("SDP value:", sdp_pl)
print("Best rounded cut:", best_pl)
print("Approx ratio (best/SDP):", best_pl / sdp_pl if sdp_pl > 1e-12 else None)
print("Best labels:", lab_pl)

# Optional: exact MAXCUT for n=10 by brute force (1024 labelings)
def exact_maxcut_bruteforce(W):
    n = W.shape[0]
    best = -1.0
    best_s = None
    for mask in range(1 << n):
        s = np.ones(n, dtype=int)
        for i in range(n):
            if (mask >> i) & 1:
                s[i] = -1
        val = cut_value_from_labels(s, W)
        if val > best:
            best = val
            best_s = s.copy()
    return best, best_s

exact_pl, exact_lab = exact_maxcut_bruteforce(W_planted)
print("Exact MAXCUT (bruteforce):", exact_pl)
print("Ratio (best rounded / exact):", best_pl / exact_pl if exact_pl > 1e-12 else None)


=== (b) Triangle K3 ===
SDP status: optimal
SDP value: 2.2500000002168905
Best rounded cut: 2.0
Best labels: [ 1 -1 -1]

=== (c) Random graph ===
Source: generated G(100,0.1), seed=0
n = 100 , |E| = 520
SDP status: optimal
SDP value: 389.6111782715055
Best rounded cut: 369.0
Approx ratio (best/SDP): 0.9470980828554608

=== (d) Planted graph ===
n = 10
SDP status: optimal
SDP value: 20999.999758632144
Best rounded cut: 21000.0
Approx ratio (best/SDP): 1.0000000114937075
Best labels: [-1 -1  1 -1  1 -1 -1  1 -1 -1]
Exact MAXCUT (bruteforce): 21000.0
Ratio (best rounded / exact): 1.0


## Question 2 SDP relaxation for MAXCUT 
### (a) SDP relaxation in CVXPY 
For an undirected weighted graph with symmetric weight matrix $W=(w_{ij})$ and $w_{ii}=0$, the MAXCUT value for a labeling $s\in\{\pm1\}^n$ is
$$
\mathrm{cut}(s)=\sum_{i<j} w_{ij}\,\frac{1-s_is_j}{2}.
$$
The standard SDP relaxation from the lectures introduces a symmetric matrix $X\in\mathbb{R}^{n\times n}$ and replaces $s_is_j$ by $X_{ij}$, with the constraints that $X$ is positive semidefinite and has unit diagonal:
$$
X\succeq 0,\qquad X_{ii}=1\ \ \forall i.
$$
The SDP is
$$
\max\ \sum_{i<j} w_{ij}\,\frac{1-X_{ij}}{2}
\quad\text{s.t.}\quad
X\succeq 0,\ \ \mathrm{diag}(X)=\mathbf{1}.
$$
Equivalently (using symmetry), this can be written as
$$
\max\ \frac14\sum_{i,j} w_{ij}(1-X_{ij})
\quad\text{s.t.}\quad
X\succeq 0,\ \ \mathrm{diag}(X)=\mathbf{1}.
$$
After solving, $X$ is a Gram matrix of unit vectors, i.e. $X_{ij}=v_i^Tv_j$ for some vectors $v_i$ with $\|v_i\|=1$. I then applied Goemans–Williamson random-hyperplane rounding: sample a random hyperplane through the origin and set $\hat s_i=\mathrm{sign}(g^Tv_i)$, repeating multiple times and keeping the best cut.



### (b) Solve the SDP for the triangle graph (from Question 1)
For the triangle $K_3$ with unit edge weights, the solver returned:
- SDP value: $2.25$ (numerically $2.2500000002168905$),
- best rounded cut value: $2.0$ with labels $[1,-1,-1]$.

This matches the theory: the true MAXCUT for a triangle is $2$, while the SDP relaxation is an upper bound and is equal to $9/4$.



### (c) Random graph from HW2 (here: generated $G(100,0.1)$ with seed $0$)
I solved the SDP on the random graph instance with:
- $n=100$, $|E|=520$,
- SDP value: $389.6111782715055$,
- best rounded cut value: $369.0$.

The approximation ratio against the SDP bound is:
$$
\frac{369.0}{389.6111782715055}\approx 0.9471.
$$
As expected, the rounded cut is a feasible cut and its value is below the SDP upper bound.



### (d) Planted partition graph $G$ (given adjacency matrix)
Using the provided adjacency matrix as $W$, the solver returned:
- SDP value: $20999.999758632144$ (approximately $21000$),
- best rounded cut value: $21000.0$,
- best labels: $[-1,-1,1,-1,1,-1,-1,1,-1,-1]$.

The approximation ratio against the SDP bound is:
$$
\frac{21000.0}{20999.999758632144}\approx 1.00000001\approx 1.
$$
Since this instance has only $n=10$, I also computed the exact MAXCUT by brute force and obtained:
- exact MAXCUT: $21000.0$.

Therefore the rounded solution is optimal for this instance:
$$
\frac{\text{best rounded}}{\text{exact MAXCUT}}=\frac{21000}{21000}=1.
$$
Overall, for this planted graph the SDP relaxation is essentially tight, and Goemans–Williamson rounding finds an optimal cut.


### Steps for implementing SDP for Maxcut

#### Formulate the Maxcut problem as SDP (recall lecture or refer to reference book [GM] in syllabus section)

####  Once you obtain 'X' (the solution positive-semi-definite matrix from SDP, you may use Cholesky decomposition to get the solution unit-vectors, stacked as column vectors in 'U')

```python
Xnew = X.value
eigs = np.linalg.eigh(Xnew)[0]
if np.min(eigs) < 0:
  Xnew = Xnew + (1.00001 * abs(min(eigs)) * np.identity(n_nodes))
elif np.min(eigs) == 0:
  Xnew = Xnew + 0.0000001 * np.identity(n_nodes)
U = np.linalg.cholesky(Xnew).T
```

#### Round the unit-vectors to appropriate partition as explained in [GM] or class.